### PR TITLE
Matteo birefnet version

### PR DESCRIPTION
**Pull request by MatteoKartoon**
Splitting, training (to fix), testing, visualization, evaluation

### CODE REVIEW

File: train_test.sh
Comment (joelkartoon) [inline train_test.sh:1]: What exactly does this do?
Seems like you're testing on always the same checkpoint

**Comment by joelkartoon** [comment]
additionally, I think it would be nice to group the scripts into a `scripts` folder for all the .py scripts, and the bash scripts in a `bash_scripts` folder

File: train.py [hunk 1]
```diff
@@ -167,36 +189,58 @@ def _train_batch(self, batch):
             gts = batch[1].to(device)
             class_labels = batch[2].to(device)
         self.optimizer.zero_grad()
-        scaled_preds, class_preds_lst = self.model(inputs)
-        if config.out_ref:
+
+        if training:
```
Comment (joelkartoon) [inline train.py:193]: I think this boolean is not clear. is this to say only whether it's `training` vs `validation`? it could easily be confused with `train` vs `test`

File: train.py [hunk 2]
```diff
@@ -167,36 +189,58 @@ def _train_batch(self, batch):
...
+        else:
+            scaled_preds = self.model(inputs)
+            class_preds_lst = None
+        
+        if config.out_ref and training:
+            # Only unpack if in training mode and out_ref is enabled
             (outs_gdt_pred, outs_gdt_label), scaled_preds = scaled_preds
             for _idx, (_gdt_pred, _gdt_label) in enumerate(zip(outs_gdt_pred, outs_gdt_label)):
-                _gdt_pred = nn.functional.interpolate(_gdt_pred, size=_gdt_label.shape[2:], mode='bilinear', align_corners=True).sigmoid()
+                _gdt_pred = nn.functional.interpolate(_gdt_pred, size=_gdt_label.shape[2:], mode='bilinear', align_corners=True)#.sigmoid()
```
Comment (joelkartoon) [inline train.py:204]: use config on whether or not we're using BCELoss
Also, is this only used for BCELoss?

File: train.py [hunk 3]
```diff
@@ -167,36 +189,58 @@ def _train_batch(self, batch):
...
                 _gdt_label = _gdt_label.sigmoid()
                 loss_gdt = self.criterion_gdt(_gdt_pred, _gdt_label) if _idx == 0 else self.criterion_gdt(_gdt_pred, _gdt_label) + loss_gdt
             # self.loss_dict['loss_gdt'] = loss_gdt.item()
-        if None in class_preds_lst:
+        if None in class_preds_lst if class_preds_lst else True:
```
Comment (joelkartoon) [inline train.py:208]: this is a somewhat confusing clause

File: train.py [hunk 4]
```diff
@@ -167,36 +189,58 @@ def _train_batch(self, batch):
...
+        
         # Loss
         loss_pix = self.pix_loss(scaled_preds, torch.clamp(gts, 0, 1)) * 1.0
         self.loss_dict['loss_pix'] = loss_pix.item()
+
         # since there may be several losses for sal, the lambdas for them (lambdas_pix) are inside the loss.py
         loss = loss_pix + loss_cls
-        if config.out_ref:
-            loss = loss + loss_gdt * 1.0
+        if config.out_ref and training:
```
Comment (joelkartoon) [inline train.py:220]: so we're disabling the `loss_gdt` for validation?

File: eval_existingOnes.py [hunk 5]
```diff
@@ -128,7 +128,7 @@ def do_eval(args):
 
     os.makedirs(args.save_dir, exist_ok=True)
     try:
-        args.model_lst = [m for m in sorted(os.listdir(args.pred_root), key=lambda x: int(x.split('epoch_')[-1]), reverse=True) if int(m.split('epoch_')[-1]) % 1 == 0]
+        args.model_lst = [m for m in sorted(os.listdir(args.pred_root), key=lambda x: int(x.split('epoch_')[-1].split('-')[0]), reverse=True) if int(m.split('epoch_')[-1].split('-')[0]) % 1 == 0]
```
[the remark this answers is not in the document]
Comment (MatteoKartoon) [reply eval_existingOnes.py:131]: this is the original version of the evaluation fine, not used anymore

File: train_test.sh
Comment (MatteoKartoon) [reply train_test.sh:1]: yes, this is a file to train and test running a single file, it was included in the original code. I tried to adapt it a bit, but I think it is unuseful for us

File: split.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+def check_correspondence(gt_images: List[str], original_images: List[str]):
+    """
+    Check if the images have been split correctly, with each annotated image coinciding with an original image
+    Return True if the splitting is correct, False otherwise
+    """
+    common_images=common_order_1(gt_images, original_images)
+    if common_images==gt_images:
+        return True
+    else:
+        print("The following images are not in the original images:")
```
[the remark this answers is not in the document]
Comment (MatteoKartoon) [reply split.py:23]: it is a way to semplify the data upload procedure: I don't have a folder containing the original version of the annotated images, so I upload the `filtred` folder and the images that gabriel discarded are ignored

File: scripts/inference.py [hunk 7]
```diff
@@ -14,16 +14,18 @@
...
+def inference(model, data_loader_test, pred_root, testset, device=0):
     model_training = model.training
     if model_training:
         model.eval()
     model.half()
 
     #get epoch from ckpt path name
     epoch = args.ckpt_path.split('_')[-1].split('.')[0]
+    #get the tr
+    training_date=args.ckpt_path.split('/')[0]
```
Comment (joelkartoon) [inline scripts/inference.py:26]: better name for variable `folder_name`

File: scripts/inference.py [hunk 8]
```diff
@@ -14,16 +14,18 @@
...
 
-def inference(model, data_loader_test, pred_root, method, testset, device=0):
+def inference(model, data_loader_test, pred_root, testset, device=0):
     model_training = model.training
     if model_training:
         model.eval()
     model.half()
 
     #get epoch from ckpt path name
     epoch = args.ckpt_path.split('_')[-1].split('.')[0]
```
Comment (joelkartoon) [inline scripts/inference.py:24]: here you should put what naming convention you're assuming

e.g `ckpt_path="20250423__1740/epoch_294.pth"`

and maybe even do assertion (that if fits the right regular expression) on the number of characters, etc

File: scripts/split.py [hunk 9]
```diff
@@ -43,35 +50,44 @@ def move_images(image_list: List[str], source_dir: str, dest_dir: str):
...
 
-def common_order_1(image_list1: List[str], image_list2: List[str]):
+def get_common_elements_of_lists(image_list1: List[str], image_list2: List[str]):
     """
     Return a list containing the common elements of the two lists, order in the same way as the first list
     """
     return [image_list1[i] for i in range(len(image_list1)) if image_list1[i] in image_list2]
 
-def split_dataset(train_ratio: float, val_ratio: float, test_ratio: float, input_dir: str, output_dir: str, gt_dir: str):
+def split_dataset(train_ratio: float, val_ratio: float, test_ratio: float, input_dir: str, output_dir: str, gt_dir: str, dataset_name: str):
```
Comment (joelkartoon) [inline scripts/split.py:90]: add `seed` as a parameter:
```
[..], seed: int = 0):
```

File: scripts/split.py [hunk 10]
```diff
@@ -92,35 +109,37 @@ def split_dataset(train_ratio: float, val_ratio: float, test_ratio: float, input
...
         print("The images have been split correctly, moving them to the right folders...")
     else:
         raise ValueError(f"Error splitting the images")
+    
+    #Rename the images to the new format
+    rename_picture_list(found_gt,images,gt_dir)
+    rename_picture_list(found_im,original_images,input_dir)
 
     # Shuffle the images to split them randomly
+    random.seed(dataset_name)
```
Comment (joelkartoon) [inline scripts/split.py:142]: use `seed`

File: scripts/train.py [hunk 11]
```diff
@@ -189,38 +194,45 @@ def _batch(self, batch, training, batch_idx):
...
                 _gdt_label = _gdt_label.sigmoid()
                 loss_gdt = self.criterion_gdt(_gdt_pred, _gdt_label) if _idx == 0 else self.criterion_gdt(_gdt_pred, _gdt_label) + loss_gdt
             # self.loss_dict['loss_gdt'] = loss_gdt.item()
-        if None in class_preds_lst if class_preds_lst else True:
+        if None in class_preds_lst:
             loss_cls = 0.
         else:
             loss_cls = self.cls_loss(class_preds_lst, class_labels) * 1.0
-            self.loss_dict['loss_cls'] = loss_cls.item()
+            if not validation:
```
Comment (joelkartoon) [inline scripts/train.py:212]:
```
loss_dict = self.loss_dict_validation else self.loss_dict_train
```

and then use this variable everywhere

File: scripts/train.py [hunk 12]
```diff
@@ -244,43 +256,41 @@ def _batch(self, batch, training, batch_idx):
...
-                self.pix_loss.lambdas_pix_last['bce'] *= 0
-                self.pix_loss.lambdas_pix_last['ssim'] *= 1
-                self.pix_loss.lambdas_pix_last['iou'] *= 0.5
-                self.pix_loss.lambdas_pix_last['mae'] *= 0.9
+            self.pix_loss.lambdas_pix_last['bce'] *= 0
+            self.pix_loss.lambdas_pix_last['ssim'] *= 1
+            self.pix_loss.lambdas_pix_last['iou'] *= 0.5
+            self.pix_loss.lambdas_pix_last['mae'] *= 0.9
+
+        if epoch==args.epochs+config.finetune_last_epochs+1:
```
Comment (joelkartoon) [inline scripts/train.py:266]: add a print / log saying you are resetting the loss moving average

**Review by joelkartoon — COMMENTED** [review]
move `models` & `evaluation` outside of `scripts`, and also only keep real scripts that have `if __name__ == "__main__":` inside that folder

File: birefnet/loss.py [hunk 13]
```diff
@@ -129,7 +129,7 @@ class PixLoss(nn.Module):
     def __init__(self):
         super(PixLoss, self).__init__()
         self.config = Config()
-        self.lambdas_pix_last = self.config.lambdas_pix_last
+        self.lambdas_pix_last = {key: self.config.lambdas_pix_last[key] * self.config.lambdas_pix_last_activated[key] for key in self.config.lambdas_pix_last}
```
Comment (joelkartoon) [inline birefnet/loss.py:132]:
```
{key: self.config.lambdas_pix_last[key] for key in self.config.lambdas_pix_last if self.config.lambdas_pix_last_activated[key] else 0.}
```

File: scripts/inference.py [hunk 14]
```diff
@@ -19,13 +20,13 @@ def inference(model, data_loader_test, pred_root, testset, device=0):
     if model_training:
         model.eval()
     model.half()
-
+    assert validate_format(args.ckpt_path), "Checkpoint path must be in the format dddddddd__dddd/epoch_ddd.pth where d is a digit"
```
Comment (joelkartoon) [inline scripts/inference.py:23]: call `validate_format` here, and do the `assert` inside the validate

File: scripts/train.py [hunk 15]
```diff
@@ -195,6 +195,10 @@ def _batch(self, batch, batch_idx, epoch_num, validation=False):
             class_labels = batch[2].to(device)
         self.optimizer.zero_grad()
         scaled_preds, class_preds_lst = self.model(inputs)
+        if validation:
```
Comment (joelkartoon) [inline scripts/train.py:198]:
```
loss_dict=self.loss_dict_validation if validation else self.loss_dict_train
```

File: scripts/train.py [hunk 16]
```diff
@@ -237,8 +232,10 @@ def _batch(self, batch, batch_idx, epoch_num, validation=False):
             if args.use_accelerate:
                 loss = loss / accelerator.gradient_accumulation_steps
                 accelerator.backward(loss)
+                #accelerator.clip_grad_norm_(self.model.parameters(), max_norm=1.0)
```
Comment (joelkartoon) [inline scripts/train.py:235]: remove commented code

File: scripts/inference.py [hunk 17]
```diff
@@ -61,7 +62,7 @@ def validate_format(value):
     Returns True if the format is valid, False otherwise
```
Comment (joelkartoon) [inline scripts/inference.py:62]: nit: this is outdated
